convex/lib/rowLevelSecurity.ts

```diff
@@ -0,0 +1,404 @@
+import {
+  DatabaseReader,
+  DatabaseWriter,
+  DocumentByInfo,
+  DocumentByName,
+  Expression,
+  FilterBuilder,
+  FunctionArgs,
+  GenericDataModel,
+  GenericTableInfo,
+  IndexRange,
+  IndexRangeBuilder,
+  Indexes,
+  MutationCtx,
+  NamedIndex,
+  NamedSearchIndex,
+  NamedTableInfo,
+  OrderedQuery,
+  PaginationOptions,
+  PaginationResult,
+  Query,
+  QueryCtx,
+  QueryInitializer,
+  SearchFilter,
+  SearchFilterBuilder,
+  SearchIndexes,
+  TableNamesInDataModel,
+} from "convex/server";
+import { GenericId } from "convex/values";
+
+export type Rule<
+  Ctx,
+  DataModel extends GenericDataModel,
+  TableName extends TableNamesInDataModel<DataModel>
+> = (
+  ctx: Ctx,
+  message: DocumentByName<DataModel, TableName>
+) => Promise<boolean>;
+
+export type Rules<Ctx, DataModel extends GenericDataModel> = {
+  [T in TableNamesInDataModel<DataModel>]?: Rule<Ctx, DataModel, T>;
+};
+
+/**
+ * Apply row level security (RLS) to queries and mutations with the returned
+ * middleware functions.
+ *
+ * Example:
+ * ```
+ * // Defined in a common file so it can be used by all queries and mutations.
+ * const {withMutationRLS} = RowLevelSecurity<{auth: Auth, db: DatabaseReader}, DataModel>(
+ *  {
+ *    cookies: async ({auth}, cookie) => !cookie.eaten,
+ *  },
+ *  {
```

Comment:
open to thoughts, but it might be clearer to take one arg: 
```
{
  read: { cookies: ...},
  write: {cookies: ...}
}
```
Also I know the motivation for making write optional vs. having a merged list, but I imagine wanting to define the rules for each table next to each other. we could do the splitting inside RowLevelSecurity so you don't have to rewrite the wrappers? e.g.:
```
{
  cookies: {
    read: ...
    write: ...
  },
  ...
}
```

Reply:
agree that is nicer